caas/kubernetes/provider/bootstrap.go

```diff
@@ -0,0 +1,758 @@
+	if err = ensureControllerConfigmapAgentConf(client, agentConfig, cleanups); err != nil {
+		return errors.Annotate(err, "creating agent config configmap for controller")
+	}
+
+	// create statefulset to ensure controller stack.
+	if err = createControllerStatefulset(client, pcfg, agentConfig, cleanups); err != nil {
+		return errors.Annotate(err, "creating statefulset for controller")
+	}
+
+	return nil
+}
+
+func createControllerStatefulset(
+	client bootstrapBroker,
+	pcfg *podcfg.ControllerPodConfig,
+	agentConfig agent.ConfigSetterWriter,
+	cleanUps *failHooks,
+) error {
+	numberOfPods := int32(1) // TODO: HA mode!
+	spec := &apps.StatefulSet{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      resourceNameGetterStatefulSet(JujuControllerStackName),
+			Labels:    stackLabelsGetter(JujuControllerStackName),
+			Namespace: client.GetCurrentNamespace(),
+		},
+		Spec: apps.StatefulSetSpec{
+			ServiceName: resourceNameGetterService(JujuControllerStackName),
+			Replicas:    &numberOfPods,
+			Selector: &v1.LabelSelector{
+				MatchLabels: stackLabelsGetter(JujuControllerStackName),
+			},
+			Template: core.PodTemplateSpec{
+				ObjectMeta: v1.ObjectMeta{
+					Labels:    stackLabelsGetter(JujuControllerStackName),
+					Namespace: client.GetCurrentNamespace(),
+				},
+				Spec: core.PodSpec{
+					RestartPolicy: core.RestartPolicyAlways,
+				},
+			},
+		},
+	}
+
+	storageclass, err := client.getDefaultStorageClass()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := buildStorageSpecForController(spec, storageclass.GetName()); err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := buildContainerSpecForController(spec, *pcfg, agentConfig); err != nil {
+		return errors.Trace(err)
+	}
+
+	logger.Debugf("creating controller statefulset: \n%+v", spec)
+	*cleanUps = append(*cleanUps, func() {
+		logger.Debugf("deleting %q statefulset", spec.Name)
+		client.deleteStatefulSet(spec.Name)
+	})
+	return errors.Trace(client.createStatefulSet(spec))
+}
+
+func buildStorageSpecForController(statefulset *apps.StatefulSet, storageClassName string) error {
+	storageSizeController, err := resource.ParseQuantity(storageSizeControllerRaw)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// build persistent volume claim.
+	statefulset.Spec.VolumeClaimTemplates = []core.PersistentVolumeClaim{
+		{
+			ObjectMeta: v1.ObjectMeta{
+				Name:   pvcNameGetterControllerPodStorage(JujuControllerStackName),
+				Labels: stackLabelsGetter(JujuControllerStackName),
+			},
+			Spec: core.PersistentVolumeClaimSpec{
+				StorageClassName: &storageClassName,
+				AccessModes:      []core.PersistentVolumeAccessMode{core.ReadWriteOnce},
+				Resources: core.ResourceRequirements{
+					Requests: core.ResourceList{
+						core.ResourceStorage: storageSizeController,
+					},
+				},
+			},
+		},
+	}
+
+	fileMode := int32(256)
+	var vols []core.Volume
+
+	// add volume log dir.
```

Comment:
We don't need a separate PV for logs though? We can use the same implementation as done for the charm operator

Reply:
yeah, removed